lib/WebSocket.js

```diff
@@ -25,277 +24,391 @@ const closeTimeout = 30 * 1000; // Allow 30 seconds to terminate the connection
 const protocolVersion = 13;
 
 /**
- * WebSocket implementation
+ * Class representing a WebSocket.
  *
- * @constructor
- * @param {String} address Connection address.
- * @param {String|Array} protocols WebSocket protocols.
- * @param {Object} options Additional connection options.
- * @api public
+ * @extends EventEmitter
  */
-function WebSocket (address, protocols, options) {
-  if (this instanceof WebSocket === false) {
-    return new WebSocket(address, protocols, options);
-  }
+class WebSocket extends EventEmitter {
+  /**
+   * Create a new `WebSocket`.
+   *
+   * @param {String} address The URL to which to connect
+   * @param {(String|String[])} protocols The subprotocols
+   * @param {Object} options Connection options
+   */
+  constructor (address, protocols, options) {
+    super();
+
+    if (typeof protocols === 'object' && !Array.isArray(protocols)) {
+      //
+      // Accept the `options` object as the 2nd argument.
+      //
+      options = protocols;
+      protocols = null;
+    }
 
-  EventEmitter.call(this);
+    if (typeof protocols === 'string') protocols = [protocols];
+    if (!Array.isArray(protocols)) protocols = [];
 
-  if (protocols && !Array.isArray(protocols) && typeof protocols === 'object') {
-    // accept the "options" Object as the 2nd argument
-    options = protocols;
-    protocols = null;
-  }
-
-  if (typeof protocols === 'string') {
-    protocols = [ protocols ];
-  }
+    this._finalize = this.finalize.bind(this);
 
-  if (!Array.isArray(protocols)) {
-    protocols = [];
-  }
+    this._binaryType = 'nodebuffer';
+    this._closeReceived = false;
+    this.bytesReceived = 0;
+    this.readyState = null;
+    this.extensions = {};
+    this._socket = null;
+    this._ultron = null;
 
-  this._socket = null;
-  this._ultron = null;
-  this._closeReceived = false;
-  this.bytesReceived = 0;
-  this.readyState = null;
-  this.extensions = {};
-  this._binaryType = 'nodebuffer';
-
-  if (Array.isArray(address)) {
-    initAsServerClient.apply(this, address.concat(options));
-  } else {
-    initAsClient.apply(this, [address, protocols, options]);
+    if (Array.isArray(address)) {
+      initAsServerClient.call(this, address[0], address[1], address[2], options);
+    } else {
+      initAsClient.call(this, address, protocols, options);
+    }
   }
-}
 
-/**
- * Inherits from EventEmitter.
- */
-util.inherits(WebSocket, EventEmitter);
+  get CONNECTING () { return WebSocket.CONNECTING; }
+  get CLOSING () { return WebSocket.CLOSING; }
+  get CLOSED () { return WebSocket.CLOSED; }
+  get OPEN () { return WebSocket.OPEN; }
 
-/**
- * Ready States
- */
-['CONNECTING', 'OPEN', 'CLOSING', 'CLOSED'].forEach(function each (state, index) {
-  WebSocket.prototype[state] = WebSocket[state] = index;
-});
+  /**
+   * @type {Number}
+   */
+  get bufferedAmount () {
+    var amount = 0;
```

Comment:
I think it's still better to keep `var` as it is consistent with the rest of the code base and we are 100% safe.
Switching to `let` should be very easy with an eslint rule when we are sure it doesn't cause any harm.